#186872583 - Instrument CLUE Performance Metrics

dennisrcao opened the pull request:
Final Log Object includes
✔️ Performance metrics (loading measurements).
 ✔️ Class/Unit/Problem/User - already existed
 ✔️ Number of docs loaded
 ✔️ Total # of tiles loaded
 ✔️ Summary of document on the right(primaryDocument) - how many tiles of each type
 ✔️ Summary of the loaded curriculum documents - how many tiles of each type
 ✔️ Measure the calculation time using `performance.now` and `console.log`.
________________________________________________________________________________________
Fulfills this ticket 

[**[Platform]** Instrument CLUE Performance Metrics
](https://www.pivotaltracker.com/n/projects/2441242/stories/186872583)

Deployed link: 
https://collaborative-learning.concord.org/branch/instrument-clue-performance-metrics/?appMode=demo&demoName=CLUE&fakeClass=1&fakeUser=student:1&problem=0.1
_____________________________________________________________

LocalHost: 
http://localhost:8080/?appMode=demo&demoName=CLUE&fakeClass=1&fakeUser=student:1&problem=0.1


Merged Boris' branch for logging messages
https://github.com/concord-consortium/collaborative-learning/pull/2163/files

dennisrcao commented on src/utilities/loading-utils.ts:
```diff
@@ -4,10 +4,17 @@ const sessionStorageStartTimeItem = 'loading-time-start';
 function getTimestamp() {
   const start = sessionStorage.getItem(sessionStorageStartTimeItem);
   if (!start) {
-    return "??";
```
I'm suggesting this get changed to `null` for cleaner type checking in `showLoadingMessage` and `removeLoadingMessage`

dennisrcao commented on src/components/document/document-workspace.tsx:
```diff
@@ -35,22 +43,90 @@ export class DocumentWorkspaceComponent extends BaseComponent<IProps> {
+    };
+
+    const endTime = performance.now();
+    console.log(`logLoadingAndDocumentMeasurements executed in ${endTime - startTime} milliseconds`);
```
Were you thinking we should add the calculation time ( `endTime - startTime`) as a property to the `finalLogObject`? Or just for us to analyze it's impact on the loading of CLUE?

scytacki replied:
I think it is fine to just put it in the console when the "loading" debug flag is on. If we find through testing that it takes a lot of time, then we are going to have to fix it or take it out.

dennisrcao commented:
I know it's failing some cypress tests and have been spending the last few hours tracking it down. 
It looks like this branch has a bug where when I hit the "Edit" button on a left side document, it won't show up on the primary workspace (rendered by `document-workspace.ts` that I made changes to). I was thinking it had to do with setting this new state I introduced so I've been comparing that with this branch off `master` called  `186872583-instrument-clue-performance-metrics-no-state`

Branch link [here](https://github.com/concord-consortium/collaborative-learning/tree/186872583-instrument-clue-performance-metrics-no-state)

scytacki reviewed (changes requested):
Yeah you shouldn't need to use state here.
Also the function `logLoadingAndDocumentMeasurements` shouldn't be in the component it should be in loading-utils. You can pass the stores into it so it has what it needs. It can also just call getLoadingMeasurements itself.

The other issue with your changes is `this.primaryDocument`. It won't be set on the first render. `componentDidMount` happens after render. I'd recommend just calling `getPrimaryDocument` both places, saving the primary document in an instance variable will cause problems when the primary document changes.

dennisrcao commented:
Implemented your changes @scytacki and re-requested review.

dennisrcao commented:
As for the rest of the TODO's: 
• Data structure needs a refactor, and we can refactor calls to sessionStorage.
↳ Insertion order should be in tact, either an array of array, set
• Detect HTTP2 or higher? - request the javascript and requesting the curriculum files ,
↳ you can look at the response and you see if it's using HTTP2.


I'm going to be doing that in parallel with the [Google Analytics Metric ticket](https://www.pivotaltracker.com/n/projects/2441242/stories/186939310), sound okay or should we break those TODO's into more seperate tickets?

scytacki commented on src/utilities/loading-utils.ts:
```diff
@@ -17,21 +30,97 @@ export function getCurrentLoadingMessage() {
+  }, 0);
+
+  const primaryDocTilesByType = primaryDocument?.content?.getAllTilesByType() as any;
+  const primaryDocNumTilesByType = countTileKeys(primaryDocTilesByType);
```
```suggestion
  const primaryDocNumTilesByType = primaryDocTilesByType ?? countTileKeys(primaryDocTilesByType);
```
This is safer since primaryDocTilesByType might be undefined based on line 79. This change might also let you remove the `as any` case on line 79.

dennisrcao replied:
The `as any` still remains, but I implemented that change.

scytacki reviewed (changes requested):
looking good. I found a few things to fix.

I think it is best to punt the remaining TODOs to a new PR. 

I'd also note that the on my chromebook the information about the documents took 150ms to generate in the CLUE demo space as a student. I think that is still an OK amount of time, but it is on the edge.

dennisrcao commented:
@scytacki I've updated this branch to reflect your changes and added the 
`TODOs` here: https://www.pivotaltracker.com/n/projects/2441242/stories/186987708

scytacki reviewed (approved):
Looks good. I'm assuming the failing tests are the same ones that are failing in master.
